rclcpp/include/rclcpp/generic_publisher.hpp

```diff
@@ -31,17 +31,34 @@
 #include "rclcpp/typesupport_helpers.hpp"
 #include "rclcpp/visibility_control.hpp"
 
+#include "rmw/rmw.h"
+
 namespace rclcpp
 {
 
+class TypeSupportInterface
+{
+protected:
+  TypeSupportInterface(
+    std::shared_ptr<rcpputils::SharedLibrary> ts_lib,
+    const std::string & topic_type)
+  : ts_lib_(ts_lib),
+    type_support_(rclcpp::get_typesupport_handle(topic_type, "rosidl_typesupport_cpp", *ts_lib)) {}
+
+  // The type support library should stay loaded, so it is stored in the GenericPublisher
+  std::shared_ptr<rcpputils::SharedLibrary> ts_lib_;
+
+  const rosidl_message_type_support_t * type_support_;
+};
+
 /// %Publisher for serialized messages whose type is not known at compile time.
 /**
  * Since the type is not known at compile time, this is not a template, and the dynamic library
  * containing type support information has to be identified and loaded based on the type name.
  *
  * It does not support intra-process handling.
  */
-class GenericPublisher : public rclcpp::PublisherBase
+class GenericPublisher : public TypeSupportInterface, public rclcpp::PublisherBase
```

Comment:
Why make this a new class and inherit from it? It doesn't appear to be used anywhere else.

And even if it is, I think we should prefer composition over inheritance in this case, i.e. have `TypeSupportInterface` be a member of `GenericPublisher` rather than a parent class to it.

Reply:
@wjwwood  

Thanks for your comments.  

> Why make this a new class and inherit from it? It doesn't appear to be used anywhere else.

In original code, rosidl_message_type_support_t is only used  in `PublisherBase::PublisherBase` and not stored. But `GenericPublisher` need this for the new interface.  
I don't want to change `PublisherBase`. So I consider changing in GenericPublisher.  

> And even if it is, I think we should prefer composition over inheritance in this case, i.e. have TypeSupportInterface be a member of GenericPublisher rather than a parent class to it.  

For this case, composition cannot be used.   
We need to save `rosidl_message_type_support_t` before calling the Constructor of PublisherBase.   
What I want as below
```c++
GenericPublisher(
    rclcpp::node_interfaces::NodeBaseInterface * node_base,
    std::shared_ptr<rcpputils::SharedLibrary> ts_lib,
    const std::string & topic_name,
    const std::string & topic_type,
    const rclcpp::QoS & qos,
    const rclcpp::PublisherOptionsWithAllocator<AllocatorT> & options)
  : type_support_handle_(rclcpp::get_typesupport_handle(topic_type, "rosidl_typesupport_cpp", *ts_lib)),
    PublisherBase(
      node_base,
      topic_name,
      *type_support_handle_,
      options.template to_rcl_publisher_options<rclcpp::SerializedMessage>(qos))
```
Actually, passed `type_support_handle_` isn't initialized by `rclcpp::get_typesupport_handle(topic_type, "rosidl_typesupport_cpp", *ts_lib)`.

Use simple code to explain
```c++
class Base {
public:
  Base(int base_data):
  base_data_(base_data)
  {
    std::cout << "Base: base_data_=" << base_data_ << std::endl;
  }
private:
  int base_data_;
};

class A : public Base {
public:
  A(int a, int b):
  a_data_(a+b),
  Base(a_data_) {
    std::cout << "A: a_data_= " << a_data_ << std::endl;
  }
private:
  int a_data_;
};

int main() {
  A a(10, 5);

  return 0;
}
```
We want to get "base_data_ == a_data_".  
In fact, output is
```
Base: base_data_=32764
A: a_data_= 15
```
So even if I use composition, I cannot pass correct `type support handle` to PublisherBase.  
I have to choose inheritance.   

I think there is another way.  
I can add a new protected variable in PublisherBase to store `rosidl_message_type_support_t`. But this will modify  PublisherBase slightly.   
What do you think ?

Reply:
Friendly ping @wjwwood

Reply:
@wjwwood can you check on this?